### PR TITLE
Add systemData handling 

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9586061

Pre-requisite for: https://github.com/Azure/ARO-RP/pull/1200 

### What this PR does / why we need it:

https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#system-metadata-for-all-azure-resources
ARM is providing us systemData field in the request headers, which is required by any new API.

We should not log any data from this or send via metrics.
We should store this into database and return data back to users when requested (PUT, PATCH, PUT)

### Test plan for issue:

Unit tests

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

File: pkg/api/v20191231preview/openshiftcluster.go
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster.go:1]: need to update admin api as well

Comment (jim-minter) [reply pkg/api/v20191231preview/openshiftcluster.go:1]: +1 on *not* retrofitting this to older APIs

File: pkg/api/openshiftcluster.go
Comment (jim-minter) [inline pkg/api/openshiftcluster.go:30]:
```suggestion
	CreatedByApplication     CreatedByType = "Application"
```
etc?

File: pkg/api/openshiftcluster.go
Comment (jim-minter) [inline pkg/api/openshiftcluster.go:39]: assuming that the ARM header is RFC compliant, use time.Time instead of date.Time.

File: pkg/api/openshiftcluster.go
Comment (jim-minter) [inline pkg/api/openshiftcluster.go:41]: `LastModifiedByType CreatedByType` jars a bit - perhaps a different name for CreatedByType ?

File: pkg/api/openshiftcluster.go
Comment (jim-minter) [inline pkg/api/openshiftcluster.go:22]: after discussion, we think that SystemData should not be a pointer, but its individual fields all should be.  See https://stackoverflow.microsoft.com/questions/242534

File: pkg/api/v20191231preview/openshiftcluster_convert.go
Comment (jim-minter) [inline pkg/api/v20191231preview/openshiftcluster_convert.go:84]: missing converter in the opposite direction, missing validation unit tests (fields should be considered immutable from the end user's perspective)

background: have tried to maintain an architectural separation so that the converters are always dumb, but there is validation that detects if a field has changed and it should not have done.

File: pkg/frontend/middleware/systemdata.go
Comment (jim-minter) [inline pkg/frontend/middleware/systemdata.go:15]: needs unit test

File: pkg/frontend/middleware/systemdata.go
Comment (jim-minter) [inline pkg/frontend/middleware/systemdata.go:24]:
```suggestion
		err := json.Unmarshal([]byte(data), &systemData)
```

File: pkg/frontend/middleware/systemdata.go
Comment (jim-minter) [inline pkg/frontend/middleware/systemdata.go:27]:
```suggestion
				log.Warnf("failed to read systemData: %v", err)
```

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch.go:46]: don't panic if r.Context().Value(middleware.ContextKeyCorrelationData) is nil

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch.go:129]: You may need a function which stomps systemData selectively.  I would call that function immediately after `doc.OpenShiftCluster.ID, doc.OpenShiftCluster.Name, doc.OpenShiftCluster.Type = oldID, oldName, oldType` - i.e. regardless of what the end user sends in PUT/PATCH, we will overwrite that data as we please

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (jim-minter) [inline pkg/frontend/openshiftcluster_putorpatch.go:124]: +1 to *not* updating the model in listCredentials/action

File: pkg/frontend/middleware/systemdata.go
Comment (mjudeikis) [inline pkg/frontend/middleware/systemdata.go:17]:
```
{"createdBy":"f707657a-xxxx-xxxx-xxxx-82704b3a99fd","createdByType":"Application","createdAt":"2021-04-20T11:36:09.3470409Z","lastModifiedBy":"f707657a-bec1-4294-9397-82704b3a99fd","lastModifiedByType":"Application","lastModifiedAt":"2021-04-20T11:36:09.3470409Z"}
```

**Comment by mjudeikis** [comment]
After this we should merge https://github.com/Azure/ARO-RP/pull/1200 soon after. It will need rebase and bit refactor

File: pkg/swagger/typewalker.go
Comment (mjudeikis) [inline pkg/swagger/typewalker.go:123]: time.Time causes this.

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (nilsanderselde) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:950]: I did a diff of this with `TestPutOrPatchOpenShiftClusterV20200430` and it looks good.

**Review by nilsanderselde — APPROVED** [review]
Looks good as is, but I do have a couple comments.

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (bennerv) [inline pkg/frontend/openshiftcluster_putorpatch.go:248]: I'm wondering if we don't want to stomp `CreatedAt`, `CreatedBy`, and `CreatedByType` as they ___should___ only be sent during the first creation by ARM, but in case they aren't the systemData for creation may be modified.  

```suggestion
 	if systemData.CreatedAt != nil && doc.OpenShiftCluster.SystemData.CreatedAt == nil {
```

Honestly might be a bad suggestion as we should only expect good behavior from ARM, just thinking though.

**Review by bennerv — COMMENTED** [review]
Looks good.  All the future stuff is confusing wrap my head around, but I understand the premise.

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (m1kola) [inline pkg/frontend/openshiftcluster_putorpatch_test.go:950]: Don't we need to update `TestPutOrPatchOpenShiftClusterV20200430` to also test for system data? I believe all API versions now handle this.

I would probably remove this test and keep only one test (`TestPutOrPatchOpenShiftCluster`) and update it with cases to handle system data. And deal with handling multiple API versions later.

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (m1kola) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:1068]: Alternativelly we can make `stompSystemData` mockable, test here that the mock is being called.

And test `stompSystemData` separately.

**Review by m1kola — CHANGES_REQUESTED** [review]
I wish ARM RPC had more details about system data, but I think this RP is in the right direction.

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (m1kola) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:950]: > I believe all API versions now handle this.

This is wrong: this PR updates only one API version. Still I'm inclined to hae only one test (at least for now): make `stompSystemData` mockable, test that we call it and test `stompSystemData` separately. as suggested in https://github.com/Azure/ARO-RP/pull/1412#discussion_r619141869.

File: pkg/api/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/openshiftcluster.go:48]: the challenge is that `time.Time` when marshaled does not comply with `omitempty` and gives you string with a lot of zeros. So hence the pointer. 
ARM provides these fields as per example in my test. So we receive partial json struct.Due to time package being bit crazy I kept all points for consistency. Add If arm for some reasons decides to send us `""` we should treat this as value we got. So again, pointers are to remove ambiguity in docs where we not sure what we will be getting.

File: pkg/api/v20210131preview/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/v20210131preview/openshiftcluster.go:228]: Will convert :)

File: pkg/frontend/frontend.go
Comment (mjudeikis) [reply pkg/frontend/frontend.go:273]: I dont see harm in having it on all methods. less probability we make a mistake and stop handing it. Data will be there for consumer to take if they need to.

File: pkg/frontend/middleware/systemdata.go
Comment (mjudeikis) [reply pkg/frontend/middleware/systemdata.go:20]: I think comment is needed :) other that that I think its overkill :)

File: pkg/frontend/middleware/systemdata.go
Comment (mjudeikis) [reply pkg/frontend/middleware/systemdata.go:28]: will change

File: pkg/frontend/middleware/systemdata_test.go
Comment (mjudeikis) [reply pkg/frontend/middleware/systemdata_test.go:22]: :shrug:  okey :)

File: pkg/frontend/middleware/systemdata_test.go
Comment (mjudeikis) [reply pkg/frontend/middleware/systemdata_test.go:70]: ack

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch.go:46]: if we failed in middleware because no data were provided or bad data were provided it will be nil. Without `, _` this call would panic. And we dont want this to happen. If systemData is nil, its nil.

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch.go:244]: Can I keep the name please ?:) I like the name :D

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch.go:248]: I think I would like to trust ARM here to do good thing and dont have "what if somebody does something bad" logic. I think We are covered from bad things with pointers already enough?

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:950]: To add to older api would mean changing api spec. And this is not allowed as this would be considered breaking change :/
So we are not retrofitting this into older apis. Old apis does not have converters into external model so we will have data in the database, but not in the external model. 

The problem with having one tests is that all the checks relies on external API model. So having full tests required support for multiple apis in the same tests. And I think it might be bit overcomplication for the test itself. Considering old API should not change we should leave the tests and not change it too. 
Or you think otherwise?

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:1068]: I was relucant to do this as this function is not guarded by any if or conditions. It should be called all the time. But yeah, maybe :)

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (nilsanderselde) [reply pkg/frontend/openshiftcluster_putorpatch.go:244]: Yes, just because of Mikalai's song :)

File: pkg/api/openshiftcluster.go
Comment (m1kola) [reply pkg/api/openshiftcluster.go:48]: `time.Time` & zeros - makes sense. I'm not convinced about `*string` and `*ActorType`. 

> Add If arm for some reasons decides to send us "" we should treat this as value we got.

I think this falls into the "what if somebody does something bad" logic category. `""` doesn't have any meaning in this context, so I think we should skip it.

File: pkg/frontend/middleware/systemdata_test.go
Comment (mjudeikis) [reply pkg/frontend/middleware/systemdata_test.go:75]: ack

File: pkg/api/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/openshiftcluster.go:48]: Will do the changes and see how they look. If code becomes abomination to read will leave as it is :) Overall agree

File: pkg/swagger/swagger.go
Comment (m1kola) [inline pkg/swagger/swagger.go:124]: As discussed on a call:
* We should ignore/remove SystemData definition generated from go struct definition
* Add a comment somewhere explaining why we ignore this (e.g. common types, etc.)

File: pkg/client/services/redhatopenshift/mgmt/2021-01-31-preview/redhatopenshift/enums.go
Comment (mjudeikis) [inline pkg/client/services/redhatopenshift/mgmt/2021-01-31-preview/redhatopenshift/enums.go:55]: This does not look right

File: pkg/swagger/typewalker.go
Comment (mjudeikis) [reply pkg/swagger/typewalker.go:123]: Should not be needed now

File: pkg/swagger/typewalker.go
Comment (mjudeikis) [inline pkg/swagger/typewalker.go:157]: Should not be needed now

File: swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/examples/OpenShiftClusters_Update.json
Comment (mjudeikis) [inline swagger/redhatopenshift/resource-manager/Microsoft.RedHatOpenShift/preview/2021-01-31-preview/examples/OpenShiftClusters_Update.json:59]: Need to populate examples

File: pkg/api/openshiftcluster.go
Comment (mjudeikis) [inline pkg/api/openshiftcluster.go:25]: I defaulted back to original type due fact common-types generates SDK with those names. So its to align and remove confusion :/

File: pkg/swagger/typewalker.go
Comment (m1kola) [reply pkg/swagger/typewalker.go:123]: > Should not be needed now

What do you mean? We need to revert this change?

File: pkg/frontend/frontend.go
Comment (mjudeikis) [reply pkg/frontend/frontend.go:273]: I dont think so. This didn't came into the review. If it will come we can add alter on :man_shrugging:

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:950]: Im still bit on the fence. For few reasons
Mocks are for interfaces. And I dont think this function qualifies to have its own interface just for this.
We gonna have more api specific tests with new api so test specific putOrPatch will be needed either way.
Having separate function will help us to make sure we dont break old apis when we move all our code to new ones (apis are not deprecatable in azure for lifetime of service).

So all this looks like a lot of effort which I dont think is required here now.

File: pkg/api/admin/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/admin/openshiftcluster.go:27]: I think this is fine for now. Data should be there for old cluster. LastUpdate one minimum.

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (mjudeikis) [reply pkg/frontend/openshiftcluster_putorpatch.go:257]: As I read and understood this - we don't need to do anything more here.
What do you mean by `roactively exclude Created* fields based on operation type (create or update).`?

File: pkg/frontend/openshiftcluster_putorpatch_test.go
Comment (m1kola) [reply pkg/frontend/openshiftcluster_putorpatch_test.go:950]: > Mocks are for interfaces. And I dont think this function qualifies to have its own interface just for this.

It definitevely doesn't deserve own interface. By mock I mean dummy/fake implementation here. I suggest that we do something similar to how we mock `time.Now()`. Here is an example ([code](https://github.com/Azure/ARO-RP/blob/1cc637ac2fb8e4707fa3f11b88f4ffaa2178fbc8/pkg/util/clientauthorizer/arm.go#L36) and [test](https://github.com/Azure/ARO-RP/blob/1cc637ac2fb8e4707fa3f11b88f4ffaa2178fbc8/pkg/util/clientauthorizer/arm_test.go)). In the same file you can see how we mock HTTP client's `Do` call.

So we in the test we will be replacing `stompSystemData` with a mock function and we will just need to check that we call it and that's it.


> We gonna have more api specific tests with new api so test specific putOrPatch will be needed either way.
> Having separate function will help us to make sure we dont break old apis when we move all our code to new ones (apis are not deprecatable in azure for lifetime of service).

We need something better than having N separate tests for each API version. Let's deal with it later. Properly.

File: pkg/api/admin/openshiftcluster.go
Comment (m1kola) [reply pkg/api/admin/openshiftcluster.go:27]: @mjudeikis as far as I understand this is not quite right. We will not have last update fields in the begining. We will only have it after receiving first PUT or PATCH request from ARM. So some old clusters which customers do not modify (from ARM perspective) will be returning `"systemData":{}`.

I think this is fine. Just want to double check.

File: pkg/frontend/openshiftcluster_putorpatch.go
Comment (m1kola) [reply pkg/frontend/openshiftcluster_putorpatch.go:257]: Discussed separately - adding for visibility.

Proxy RP always receives `Created*` fields along with `LastModified*` and should decide whether to use `Created*` values or not (this is what I meant by proactively exlucindg).

This however doesn't apply to ARO RP.

File: pkg/api/admin/openshiftcluster.go
Comment (mjudeikis) [reply pkg/api/admin/openshiftcluster.go:27]: Where do they will return it? Swagger does not have this field in old API, nor in our side old APIS?
This will get returned only if customer interact with old cluster using new api. And for this I think this is fine to have empty values.

Comment (m1kola) [reply pkg/api/admin/openshiftcluster.go:27]: > This will get returned only if customer interact with old cluster using new api. And for this I think this is fine to have empty values.

Yes, this is exactly what I mean: `GET`ing an old cluster via new API before making any `PUT`s or `PATCH`es.

I also think it is fine. Again: just wanted to double check.

**Review by m1kola — APPROVED** [review]
Looks good to me

**Pull request merged**
